pkg/compute/deploy.go

```diff
@@ -41,137 +56,437 @@ func NewDeployCommand(parent common.Registerer, client api.HTTPClient, globals *
+		// and the associated domain/backend and store the Service ID within the
+		// manifest. On subsequent runs of the deploy subcommand we'll skip the
+		// service/domain/backend creation.
+		//
+		// NOTE: we're shadowing the `version` and `serviceID` variable.
+		version, serviceID, err = createService(progress, c.Globals.Client, name, desc)
+		if err != nil {
+			return err
 		}
 
-		c.Path = filepath.Join("pkg", fmt.Sprintf("%s.tar.gz", sanitize.BaseName(name)))
-	}
+		undoStack.Push(func() error {
+			return c.Globals.Client.DeleteService(&fastly.DeleteServiceInput{
+				ID: serviceID,
+			})
+		})
 
-	serviceID, source := c.manifest.ServiceID()
-	if source == manifest.SourceUndefined {
-		return fmt.Errorf("error reading service: no service ID found. Please provide one via the --service-id flag or within your package manifest")
-	}
+		// NOTE: We have to mark the progress as 'done' before prompting the user
+		// for input because otherwise the progress will overwrite the input
+		// prompt. This means after we've prompted the user for input we recreate a
+		// progress instance.
+		progress.Done()
 
-	// Set the version we want to operate on.
-	// If version not provided infer the latest ideal version from the service.
-	if !c.Version.WasSet {
-		progress.Step("Fetching latest version...")
-		versions, err := c.Globals.Client.ListVersions(&fastly.ListVersionsInput{
-			ServiceID: serviceID,
-		})
+		domain, err := cfgDomain(c.Domain, defaultTopLevelDomain, out, in, validateDomain)
 		if err != nil {
-			return fmt.Errorf("error listing service versions: %w", err)
+			return err
 		}
 
-		version, err = getLatestIdealVersion(versions)
+		backend, backendPort, err := cfgBackend(c.Backend, c.BackendPort, out, in, validateBackend)
 		if err != nil {
-			return fmt.Errorf("error finding latest service version")
+			return err
+		}
+
+		progress = newProgress(c.Globals.Verbose(), out)
+
+		err = createDomain(progress, c.Globals.Client, serviceID, version.Number, domain, undoStack)
+		if err != nil {
+			return err
+		}
+
+		err = createBackend(progress, c.Globals.Client, serviceID, version.Number, backend, backendPort, undoStack)
+		if err != nil {
+			return err
+		}
+	}
+
+	// We only want to update the manifest if it's the first time deploying and
+	// so we will have only just created a new service. If the Service ID is
+	// provided by the flag and not the file, then we'll also store that ID
+	// within the manifest.
+	if source == manifest.SourceUndefined || source != manifest.SourceFile {
+		err = updateManifestServiceID(ManifestFilename, progress, serviceID)
+		if err != nil {
+			return err
 		}
-	} else {
-		version = &fastly.Version{Number: c.Version.Value}
 	}
 
 	progress.Step("Validating package...")
+	if err := validate(path); err != nil {
+		return err
+	}
+
+	cont, err := pkgCompare(c.Globals.Client, serviceID, version.Number, path, progress, out)
+	if err != nil {
+		return err
+	}
+	if !cont {
```

Comment:
9. We'll short-circuit the `deploy` subcommand if we discover the actual wasm binary hasn't been changed since the last time it was uploaded.